storage/src/bmt/mod.rs

```diff
@@ -0,0 +1,809 @@
+    hasher: H,
+    leaves: Vec<H::Digest>,
+}
+
+impl<H: Hasher> Builder<H> {
+    /// Creates a new Binary Merkle Tree builder.
+    pub fn new(leaves: usize) -> Self {
+        Self {
+            hasher: H::new(),
+            leaves: Vec::with_capacity(leaves),
+        }
+    }
+
+    /// Adds a leaf to the Binary Merkle Tree.
+    ///
+    /// When added, the leaf is hashed with its position.
+    pub fn add(&mut self, leaf: &H::Digest) {
+        let position: u32 = self.leaves.len().try_into().expect("too many leaves");
+        self.hasher.update(&position.to_be_bytes());
+        self.hasher.update(leaf);
+        self.leaves.push(self.hasher.finalize());
+    }
+
+    /// Builds the Binary Merkle Tree.
+    pub fn build(self) -> Option<Tree<H>> {
+        Tree::new(self.hasher, self.leaves)
+    }
+}
+
+/// Constructed Binary Merkle Tree (BMT).
+#[derive(Clone, Debug)]
+pub struct Tree<H: Hasher> {
+    /// The digests at each level of the tree (from leaves to root).
+    levels: Vec<Vec<H::Digest>>,
+}
+
+impl<H: Hasher> Tree<H> {
+    /// Builds a Merkle Tree from a slice of position-hashed leaf digests.
+    ///
+    /// If `leaves` is empty, returns `None`.
+    fn new(mut hasher: H, leaves: Vec<H::Digest>) -> Option<Self> {
+        // Ensure there are non-zero leaves
+        if leaves.is_empty() {
+            return None;
+        }
+
+        // Create the first level
+        let mut levels = Vec::new();
+        levels.push(leaves);
+
+        // Construct the tree level-by-level
+        let mut current_level = levels.last().unwrap();
+        while current_level.len() > 1 {
+            let mut next_level = Vec::with_capacity((current_level.len() + 1) / 2);
+            for chunk in current_level.chunks(2) {
+                // Hash the left child
+                hasher.update(&chunk[0]);
+
+                // Hash the right child
+                if chunk.len() == 2 {
+                    hasher.update(&chunk[1])
+                } else {
+                    // If no right child exists, duplicate left child.
+                    hasher.update(&chunk[0]);
+                };
+
+                // Compute the parent digest
+                next_level.push(hasher.finalize());
+            }
+
+            // Add the computed level to the tree
+            levels.push(next_level);
+            current_level = levels.last().unwrap();
+        }
+        Some(Self { levels })
+    }
+
+    /// Returns the root of the tree.
+    pub fn root(&self) -> H::Digest {
+        self.levels.last().unwrap().first().unwrap().clone()
+    }
+
+    /// Generates a Merkle proof for the leaf at `position`.
+    ///
+    /// The proof contains the sibling digest at each level needed to reconstruct
+    /// the root.
+    pub fn proof(&self, position: u32) -> Option<Proof<H>> {
+        // Ensure the position is within bounds
+        if position >= self.levels.first()?.len() as u32 {
+            return None;
```

Comment:
I guess I still don't fully understand when Option is appropriate vs Result, but this feels like it should be a Result (if not panic)? (Proving an invalid position seems like a user error.)

Reply:
That is definitely more rust-y. I'll make that change for both build/proof.

Reply:
done!